Move volume scheduling and local storage to beta

msau42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
* Move the feature gates and APIs for volume scheduling and local storage to beta
* Update tests to use the beta fields
@kubernetes/sig-storage-pr-reviews 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59390

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
ACTION REQUIRED: VolumeScheduling and LocalPersistentVolume features are beta and enabled by default.  The PersistentVolume NodeAffinity alpha annotation is deprecated and will be removed in a future release.
```

msau42 committed:
Add VolumeNodeAffinity to PersistentVolumeSpec

msau42 committed:
Add new volume-scheduler cluster role to scheduler

msau42 committed:
Volume node affinity enforcement

msau42 committed:
Generated files

lichuqiang commented:
/cc

msau42 commented:
Local PV E2Es are passing

msau42 commented:
/assign @verult @jsafrane

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -237,6 +237,13 @@ func GetClassForVolume(kubeClient clientset.Interface, pv *v1.PersistentVolume)
 // CheckNodeAffinity looks at the PV node affinity, and checks if the node has the same corresponding labels
 // This ensures that we don't mount a volume that doesn't belong to this node
 func CheckNodeAffinity(pv *v1.PersistentVolume, nodeLabels map[string]string) error {
+	if err := checkAlphaNodeAffinity(pv, nodeLabels); err != nil {
```
@jsafrane for storageclass, did the alpha annotation override beta?  Or the other way around?

msau42 replied:
Actually this is different, it is evaluating both if they both exist.

msau42 commented:
TODO: need to invalidate equivalence cache for pv update of nodeaffinity field, like in #59335

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -141,7 +141,7 @@ var (
 		Level: "s0:c0,c1"}
 )
 
-var _ = utils.SIGDescribe("PersistentVolumes-local [Feature:LocalPersistentVolumes] [Serial]", func() {
```
I may still want to keep the Feature tag.  To test GCE local SSD, we need to bring up a test cluster with a special deployment flag, so I'm thinking that it would be better to have a different test suite.  But then we would also potentially miss out on stress testing by running in parallel with other CI tests.

verult replied:
If I understand correctly, the deciding factor is whether we want to add local SSDs to normal CI clusters, and if we don't, we'd want to replicate some of the stress testing in the new cluster to ensure better coverage. This might be a question test-infra could answer.

msau42 replied:
Yup, I'll check with test-infra if they're ok with that, thanks!

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -467,6 +467,16 @@ type PersistentVolumeSpec struct {
+// VolumeNodeAffinity defines constraints that limit what nodes this volume can be accessed from.
+type VolumeNodeAffinity struct {
+	// Required specifies hard node constraints that must be met.
+	Required *NodeSelector
```
I did not follow all the discussion closely. Why it changed from `RequiredDuringSchedulingIgnoredDuringExecution` to simple `Required`? What happens when admin edits the selector when a pod is already scheduled / running? Should it be allowed?

msau42 replied:
I think the "RequiredDuringExecution" clause doesn't make sense for volumes because:
- That means the storage backend made the volume inaccessible on that node while it was still being used?  I don't think any sane backend would do that and cause data corruption for the application.
- Maybe the storage backend migrated the data to another node and but it's still globally accessible.  Then that becomes a preference instead of a requirement.  Also, I don't see a good reason why a backend would do that because it decrease workload performance.

In that case, maybe it is good to disallow modifying the Required field.  I can envision future scenarios for allowing changes to a Preferred field though.

thockin replied:
In general, disallow field updates until/unless you know exactly what they are for.

msau42 replied:
Yup, it disallows updates now.  Hm but actually this will make it difficult for users to upgrade from alpha -> beta, or for automatically upgrading zonal PVs like GCE PDs to the beta format.

I think what I need to do is:
- Allow setting NodeAffinity if it was previously non-nil
- Disallow modifying NodeAffinity once its set.

msau42 commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -639,12 +588,142 @@ func TestValidateLocalVolumes(t *testing.T) {
 					StorageClassName: "test-storage-class",
 				}),
 		},
+		"valid local volume": {
```
Done

luxas reviewed (commented):
Are we confident enough in this feature to enable it by default or should we graduate it to beta stability level first and then go GA with it (and enable by default) in the next release?

Just asking, as I haven't followed this feature very closely but am curious.

msau42 commented:
Hi @luxas, even if the feature is turned on by default, it is still opt-in to use the new functionality.  In order to use the feature, you need to explicitly set certain API fields to non-default values, so no existing workloads will have changing behavior due to this feature.

In terms of confidence, there are already extensive unit, integration, and e2e tests that have been running since 1.7.  What is missing are longevity tests to catch things like memory leaks.  I can definitely add some stress tests and then manually verify resource usage.  I'm not aware of anything in our test framework that can automate something like this though.

msau42 commented:
/test pull-kubernetes-bazel-test

msau42 commented:
/retest

jsafrane commented:
LGTM from my point of view

verult commented:
/lgtm

msau42 commented:
/assign @thockin @liggitt

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -483,15 +483,13 @@ func ClusterRoles() []rbac.ClusterRole {
-			}
-		}
+		roles = append(roles, rbac.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{Name: "system:kube-volume-scheduler"},
```
nit - maybe just `system:volume-scheduler`, since this no longer matches the component name, and the role describes permissions to schedule volumes?

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -483,15 +483,13 @@ func ClusterRoles() []rbac.ClusterRole {
+		roles = append(roles, rbac.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{Name: "system:kube-volume-scheduler"},
+			Rules: []rbac.PolicyRule{
+				rbac.NewRule("update").Groups(legacyGroup).Resources("persistentvolumes").RuleOrDie(),
```
include `get`,`list`,`watch`, `update`, `patch`, to make this role sufficient for all the read/update operations you'd need to do on a PV

msau42 replied:
done

luxas commented:
Thanks @msau42 for the information!

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4958,6 +4969,27 @@ func validateStorageNodeAffinityAnnotation(annotations map[string]string, fldPat
+	}
+
+	if nodeAffinity.Required != nil {
+		allErrs = append(allErrs, ValidateNodeSelector(nodeAffinity.Required, fldPath.Child("required"))...)
```
I'm in the middle of designing dynamic provisioning with topology and looking at how that's going to work with CSI.  The way that topology is specified in the CSI spec is not going to be able to support the full functionality of a NodeSelector.  The CSI spec represents topology as a map<string>[]string, where the key is the label key, and the value is an array of label values.  This is sufficient to support topologies of all the storage systems I can think of.

But I think from a Kubernetes standpoint, there is value in adopting consistent structs to describe topology across the system, even if it can't be fully supported.  So I still prefer representing volume topology with NodeAffinity in the Kubernetes API, and I could add validation here to only allow for the "In" operator.

@thockin @saad-ali @jsafrane and anyone else, would appreciate your thoughts

thockin replied:
Which semantics are not expressible?  Is CSI just "equals" or is it "in set".  What are the gaps?  "not in"?  What else?

What facet of CSI is this touching?  It's been a while since I looked at CSI in depth, but this starts to feel like it might be scope creep.

msau42 replied:
Basically all the other node selector operators don't really make sense to describe volume accessibility: "NotIn", "Exists", "DoesNotExist", "Gt", "Lt".

For just the specification of volume topology + scheduling of those volumes (which is this PR), CSI can support this (the csi-kubernetes adapter can translate from csi topology -> k8s volume topology).  But once we consider dynamic provisioning, we need to do the translation the other way (k8s topology -> csi topology for the csi volume create call).  Agree that the CSI problem doesn't directly impact this PR, however, any new topology APIs we may have to add for dynamic provisioning could be inconsistent with how topology is specified here.

thockin replied:
I can make up use-cases for "not-in", but they can probably be solved by reciprocal "in".

Can you walk me through a hypothetical where we might use parts of NodeAffinity that CSI can't halndle?

msau42 replied:
We discussed the use case of restricting topology in the StorageClass, moving it out from a plugin-specific parameter to a first class field, and then having to pass it to the provisioners.  In that use case, a NodeSelector doesn't make sense, and it's ok for the topology specification to be different as long as the underlying foundation of specifying topology keys and using node labels is similar.

So conclusion is: PV.NodeAffinity is fine the way it is.  Topology for dynamic provisioning may be more limited and can be specified using different APIs (and will be initially alpha).

msau42 replied:
If no one else has concerns, then this PR is ready to go.

jsafrane commented:
/lgtm
from storage point of view

msau42 commented:
/assign @thockin for approval

msau42 commented:
I rebased and also added a commit to update volume predicate invalidation for the new NodeAffinity field.

/assign @resouer @bsalamat 
for scheduler review

msau42 commented:
Nevermind, ignore my last comment.  There is validation that the PV.NodeAffinity field cannot be changed.  So there is no need to update predicate invalidation for PV update.  I removed the last commit.

jsafrane commented:
/lgtm

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1497,6 +1500,10 @@ func ValidatePersistentVolume(pv *core.PersistentVolume) field.ErrorList {
 	nodeAffinitySpecified, errs := validateStorageNodeAffinityAnnotation(pv.ObjectMeta.Annotations, metaPath.Child("annotations"))
 	allErrs = append(allErrs, errs...)
 
+	volumeNodeAffinitySpecified, errs := validateVolumeNodeAffinity(pv.Spec.NodeAffinity, specPath.Child("nodeAffinity"))
```
What if both this field and the older annotation are set?

msau42 replied:
The current code will evaluate both if both are set, but it can be changed to have beta be ignored if alpha is set.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4952,6 +4963,27 @@ func validateStorageNodeAffinityAnnotation(annotations map[string]string, fldPat
 	return policySpecified, allErrs
 }
 
+// validateVolumeNodeAffinity tests that the PersistentVolume.NodeAffinity has valid data
```
document the bool return

thockin reviewed (commented):
Please clarify and handle the transitional case where both are set.  The alpha should win.

msau42 committed:
Disallow setting both alpha and beta PV nodeAffinity
Allow setting PV nodeAffinity if previously unset

thockin commented:
/lgtm
/approve

ryanwalls commented:
Is there documentation for how to use persistent local storage?  Can't seem to find anything.

mightwork commented:
@ryanwalls [https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume](url)

ryanwalls commented:
@mightwork FYI, get a 404 on that link.

mightwork commented:
@ryanwalls my bad. Updated with the correct link.
